src/main/java/ml/shifu/shifu/core/dtrain/dt/DTEarlyStopDecider.java

```diff
@@ -0,0 +1,307 @@
+    /**
+     * trainErrorDecider return a positive signal or negative sign whether it worth further loop according to the
+     * current iteration gain.
+     */
+    private MinAverageDecider trainErrorDecider;
+
+    /**
+     * validationErrorDecider return a positive or negative sign whether training is over fitted.
+     */
+    private MinAverageDecider validationErrorDecider;
+    /**
+     * Continue count of positive sign not worth further iteration
+     */
+    private int trainGainContinueLowerCount;
+    /**
+     * Continue count of positive sign over fit
+     */
+    private int validationGainContinueNegCount;
+
+    public DTEarlyStopDecider(int treeDepth) {
+        if (treeDepth <= 0) {
+            throw new IllegalArgumentException("Tree num should not be less or equal than zero!");
+        }
+
+        this.treeDepth = treeDepth;
+
+        trainErrorDecider = new MinAverageDecider(this.treeDepth, this.treeDepth * this.magicNumber) {
+            @Override
+            public boolean getDecide() {
+                return this.gain < this.maxGain / 1000;
+            }
+        };
+
+        validationErrorDecider = new MinAverageDecider(this.treeDepth, this.treeDepth * this.magicNumber) {
+            @Override
+            public boolean getDecide() {
+                return this.gain < 0;
+            }
+        };
+    }
+
+    /**
+     * Add new iteration's train error and validation error into the decider.
+     *
+     * @param trainError training error
+     * @param validationError validation error
+     * @return true if no more iteration needed, else false
+     */
+    public boolean add(double trainError, double validationError) {
+        boolean trainErrorDecideReady = this.trainErrorDecider.add(trainError);
+        boolean validationDecideReady = this.validationErrorDecider.add(validationError);
+
+        if (trainErrorDecideReady) {
+            if(trainErrorDecider.getDecide()){
+                this.trainGainContinueLowerCount += 1;
+                LOG.warn("Continue " + this.trainGainContinueLowerCount + " positive sign for not worth more iteration!");
+            } else {
+                this.trainGainContinueLowerCount = 0;
+            }
+        }
+
+        if (validationDecideReady) {
+            if(validationErrorDecider.getDecide()){
+                this.validationGainContinueNegCount += 1;
+                LOG.warn("Continue " + this.validationGainContinueNegCount + " positive sign for not worth more iteration!");
+            } else {
+                this.validationGainContinueNegCount = 0;
+            }
+        }
+
+        return canStop();
+    }
+
+    private boolean canStop() {
+        if (this.validationGainContinueNegCount >= magicNumber) {
+            return true;
+        }
+        if (this.trainGainContinueLowerCount >= magicNumber) {
+            return true;
+        }
+        return false;
+    }
+
+    abstract class MinAverageDecider {
```

Comment:
Make it as static please, static is better than non-static in nested class, please read reference from Effective Java